docs/standards/javascript.md

```diff
@@ -0,0 +1,203 @@
+
+## Accessing CiviCRM API from JS
+
+If the current user has sufficient permissions (usually "Access CiviCRM") then your script can call the api and accomplish almost anything in CiviCRM. The syntax is:
+```javascript
+CRM.api('entity', 'action', {params}, {success: function}).
+```
+For more details, see [AJAX API](http://wiki.civicrm.org/confluence/display/CRMDOC43/AJAX+Interface) docs.
+
+## Server-Side Variables
+
+Just as you can dynamically add scripts to the page you can also add variables. They will be added to the CRM object in JSON format. Because this object is shared by many CiviCRM components, you should choose a unique namespace (typically the name of your extension). Example:
+
+```php
+// On the Server:
+CRM_Core_Resources::singleton()->addVars('myNamespace', array('foo' => 'bar'));
+```
+
+```javascript
+// In the JS file or inline script
+CRM.alert(CRM.vars.myNamespace.foo); // Alerts "bar"
+```
+
+## Localization
+
+As with PHP coding, any string that will be displayed to the user should be wrapped in ts() to translate the string. e.g. ts('hello')
+
+When your script file is being added to the page by CRM_Core_Resources it will be automatically scanned for all strings enclosed in ts() and their translations will be added as client-side variables. The javascript ts() function will use these localized strings. It can also perform variable substitution. See example below.
+
+!!!Note
+
+Because translated strings are added to the client-side by CRM_Core_Resources addScriptFile method, they will not be automatically added if you include your js on the page any other way. The ts() function will still work and perform variable substitution, but the localized string will not be available. There are 3 solutions to this problem depending on your context:
+
+1. If this is an inline script in a smarty template, use the {ts} function (see legacy issues below for an example). Note that {ts} cannot handle client-side variable substitution, only server-side template variables.
+2. If this is an inline script in a php file, use the php ts() function to do your translation and concatenate the result into your script. Or, if you need client-side variable substitution use the 3rd solution:
+3. If this is a javascript file being added to the page in a nonstandard way (or is one of the above two scenarios but you need client-side variable substitution), you could manually add any needed strings using the CRM_Core_Resources addString method
+
+## UI Elements
+
+CiviCRM ships with a number of UI widgets and plugins to create standardized "look and feel" More information can be found in [UI Elements Reference](https://wiki.civicrm.org/confluence/display/CRMDOC/UI+Elements+Reference).
+
+## Quality Control
+
+CiviCRM has a number of mechanisms that it uses to ensure quality of javascript that is submitted to core.
+
+### JSHint
+
```

Comment:
@totten @seanmadsen i copied this across from Wiki but thinking that because we have a section up the top on coding standards I think perhaps we loose this part on JSHint and change this section to be test tools perhaps?